pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -395,17 +395,32 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
         ctx.close();
     }
 
-    /*
-     * If authentication and authorization is enabled(and not sasl) and
-     *  if the authRole is one of proxyRoles we want to enforce
-     * - the originalPrincipal is given while connecting
-     * - originalPrincipal is not blank
-     * - originalPrincipal is not a proxy principal
+    /**
+     * When transitioning from Connecting to Connected, this class validates that the {@link #authRole} and the
+     * {@link #originalPrincipal} are a valid combination. Valid combinations fulfill the following rule:
+     * <p>
+     * The {@link #authRole} is in {@link #proxyRoles}, if, and only if, the {@link #originalPrincipal} is set to a role
+     * that is not also in {@link #proxyRoles}.
      */
-    private boolean invalidOriginalPrincipal(String originalPrincipal) {
-        return (service.isAuthenticationEnabled() && service.isAuthorizationEnabled()
-                && proxyRoles.contains(authRole) && (StringUtils.isBlank(originalPrincipal)
-                || proxyRoles.contains(originalPrincipal)));
+    private void validateRoleAndOriginalPrincipal() {
+        String errorMsg = null;
+        if (proxyRoles.contains(authRole)) {
+            if (StringUtils.isBlank(originalPrincipal)) {
+                errorMsg = "originalPrincipal must be provided when connecting with a proxy role.";
+            } else if (proxyRoles.contains(originalPrincipal)) {
+                errorMsg = "originalPrincipal cannot be a proxy role.";
+            }
+        } else if (StringUtils.isNotBlank(originalPrincipal)) {
+            errorMsg = "cannot specify originalPrincipal when connecting without valid proxy role.";
+        }
+        if (errorMsg != null) {
+            service.getPulsarStats().recordConnectionCreateFail();
```

Comment:
what about throwing an exception in order to let the caller code to catch it? 
In this case there's the exactly same code in case of Exception
https://github.com/apache/pulsar/blob/a0e603a3eaf822cb0c41f292ad5845ee54df0ccf/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java#L968-L973

Reply:
I agreed with @nicoloboschi.

Reply:
I considered this, but I decided not to because throwing an exception is expensive and at this point in the `ServerCnx` lifecycle, I think we should avoid using exceptions as control flow when possible.

Reply:
I extracted the code into a shared method so we can avoid an exception and avoid the code duplication.